### PR TITLE
Expose GetFreePorts from testutils

### PR DESCRIPTION
**Pull request by peter-edge**
This exposes the GetFreePorts function from testutils. This also makes it so that once you have found a free port, it is locked down, including across processes, until you release it. This will be useful when I want to parallelize the testing of examples, and when we have more integration testing that does client/server interaction.

### CODE REVIEW

**Comment by kriskowal** [comment]
I’m a skeptic on this one. I favor listening on 0 and communicating the assigned address. When is that insufficient?

**Comment by peter-edge** [comment]
A lot of our code needs an address, not a listener, and you need to share that between the client and server. You can't pass in `localhost:0` to the server and then just know the port to use for the client, our existing code doesn't expose such things. Note this function was already being used, this just exposes it and adds a locking scheme.

**Comment by kriskowal** [comment]
Yeah, I’m proposing that the right thing to do is to coordinate like:

1. everyone listen
2. coordinate
3. everyone connect

I’m in favor of adjusting whatever needs adjustment to promote this as the norm. We have been burned before by listen races. That we’ve been using this isn’t a compelling reason to continue using it.

**Comment by peter-edge** [comment]
Ya that's fine, we can talk about it, in the mean time I need something for testing though :) I can put up a proposal for the better way you just proposed though.

**Comment by peter-edge** [comment]
@kriskowal I also want to use this in service-test because I want to be able to allocate a port, and then pass it to separate client and server binaries - there is DEFINITELY no way to do this with how the examples are written now, other than to use this function. So I'll want this functionality regardless.

**Comment by kriskowal** [comment]
@peter-edge First, thank you for your initiative. Second, anyone can override me on this feedback. You have my apology for the round-trip time, and I will try to grease the wheels for anything I can earnestly green stamp or correct after review. However, my experience is that it is much harder to walk backward than forward.

Getting "free" ports is not technically sound because of the race between when you close the port to make it available to the child process and when the child process attempts to listen. In short, this introduces flappiness. We had such a function in tchannel-node two years ago and had to pull it out to fix flappiness.

Thank you for clarifying that this is for tests. I see a few ways forward.

1. We can land this and use it for examples only. This is less than idea because it will flap. It is also less than ideal because it provides an example that should not be imitated. It will lead a proliferation of a concurrency hazard.
2. We can live with the compromise that examples use static ports. This is less than ideal because we want to test the examples.
3. We can thread a named pipe instead of a port. This is less than ideal because it requires us to boil the ocean to get support for all tested transports.

I am in favor of option 2.

**Comment by breerly** [comment]
I think I'm with @kriskowal on this one - it's not great, but it's what we've got for the moment.

**Comment by peter-edge** [comment]
I'd propose the following - we merge this or some variation, but we file an issue to tackle this problem, and as part of my (or others') learning the codebase, I'll think about a better solution, that may or may not introduce api changes (ie we may have new methods that mean we want to deprecate existing methods for 2.0). How does that sound?

**Comment by breerly** [comment]
@peter-edge hmmm.... do you think its worth the risk of introducing races?

I've worked on enough codebases with pesky races that I really want to avoid that...

We can trust you'll make this work, but you'll really have to own the problem - are you up for that?

**Comment by peter-edge** [comment]
I want to use this for examples testing (where we currently have
overlapping ports that make it not possible to do proper parallel example
testing), and integration tests that use the examples (only protobuf does
this now, and it's effectively already using this code, it's surfaced via
public methods that call the current getFreePort() function).

Let's talk on Monday - I would be willing to own this, but I want to make
sure I fully understand the problem and associated issues with it.

File: internal/testutils/testutils.go [hunk 1]
```diff
@@ -23,8 +23,12 @@ package testutils
 import (
 	"fmt"
 	"net"
+	"os"
+	"path/filepath"
 	"strconv"
 
+	"github.com/nightlyone/lockfile"
```
Comment (willhug) [inline internal/testutils/testutils.go]: import order

**Comment by kriskowal** [comment]
I asked our Mesos folks what placement systems do to solve this problem. It’s way less elaborate than what I expected. They reserve a block of high but not ephemeral ports and assign them down to tasks, communicating their numbers with ephemeral ports.

While imperfect, a similar solution for tests might be viable, and less prone to races than listening on port 0 and closing. To reduce the probability of a conflict, we could start at any port between 16384–32768 and assign ports sequentially from that number.

Alternatively we can make use of SO_REUSEPORT, listening *and keeping open* the socket on orchestrating process.

**Pull request closed without merging**